# Исследование надёжности заёмщиков

Целью настоящего исследования является выявление факторов, влияющих на факт погашения кредита в срок. Результаты исследования будут использованы для построения модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Задача исследования — изучить полученную от банка статистику о платёжеспособности клиентов и ответить на следующие вопросы:
1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Знакомство с данными</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Знакомство с данными

In [1]:
#импорт библиотек
import pandas as pd
from pymystem3 import Mystem
from collections import Counter

In [2]:
df = pd.read_csv('/datasets/data.csv') #открытие файла

In [3]:
df.head(10) #для знакомства с данными посмотрим на первые и последние 10 строк датафрейма

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
df.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

### Вывод

Статистика о платежеспособности клиентов была предоставлена заказчиком в виде файла в формате CSV. Методами библиотеки pandas мы прочитали файл и вызвали 10 строк из начала и конца таблицы. Из них уже видны некоторые ошибки, которые могут повлиять на точность нашего анализа: отрицательные значения в столбце "общий трудовой стаж в днях", разный регистр текста в столбце "уровень образования клиента". Среди целей можно отметить, например, "сыграть свадьбу" и "на проведение свадьбы", которые являются по смыслу одной и той же целью, но программой будут засчитаны как две разные, что тоже повлияет на данные на выходе. Поэтому для повышения точности результатов анализа данных нам необходимо их предварительно обработать.

## Предобработка данных

### Обработка пропусков

In [5]:
df.info() #посмотрим на свойства таблицы с данными

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Как видно из краткой сводки, в столбцах days_employed и total_income есть пропущенные значения (в них по 19351 записей при общем количестве 21525). Оба столбца содержат числовые данные типа float64.

In [6]:
df[df['days_employed'].isna()].sample(10) #найдем все строки с пропусками в days_employed и посмотрим на случайные 10

children  days_employed  dob_years education  education_id  \
6725          0            NaN         57   среднее             1   
8153          0            NaN         38   Среднее             1   
2662          0            NaN         23   среднее             1   
16019         3            NaN         29   среднее             1   
17291         0            NaN         56   среднее             1   
12545         1            NaN         46    высшее             0   
19071         1            NaN         36   среднее             1   
11453         2            NaN         35   среднее             1   
12793         0            NaN         50    высшее             0   
8015          0            NaN         58   среднее             1   

               family_status  family_status_id gender income_type  debt  \
6725         женат / замужем                 0      F   пенсионер     1   
8153         женат / замужем                 0      F   сотрудник     0   
2662   Не женат / не замужем                 4      M   сотрудник     0   
16019       гражданский брак                 1      M   сотрудник     0   
17291       гражданский брак                 1      F   сотрудник     0   
12545        женат / замужем                 0      M   сотрудник     0   
19071              в разводе                 3      M   сотрудник     0   
11453       гражданский брак                 1      F   компаньон     1   
12793       гражданский брак                 1      F   сотрудник     0   
8015         женат / замужем                 0      M   сотрудник     0   

       total_income                                purpose  
6725            NaN                покупка жилья для сдачи  
8153            NaN                      операции с жильем  
2662            NaN                      операции с жильем  
16019           NaN                   покупка недвижимости  
17291           NaN                  на проведение свадьбы  
12545           NaN                покупка жилья для сдачи  
19071           NaN                          покупка жилья  
11453           NaN  получение дополнительного образования  
12793           NaN                                свадьба  
8015            NaN                   покупка недвижимости

Судя по таблице выше, где пропущено значение дней трудового стажа, там и в графе "ежемесячный доход" значится NaN, и количество пропусков в обоих столбцах совпадает. Можно предположить, что некоторые клиенты банка по каким-то причинам не стали сообщать величину трудового стажа и дохода, когда подавали заявку на кредит. 

Семейное положение, образование, род занятий и цели заёмщиков, по которым недостаёт данных, разнообразны. Если просто удалить строки с пропусками, мы потеряем важную информацию из других столбцов. Воспользуемся методом fillna() и заменим отсутствующие значения в столбце days_employed нулями.

In [7]:
df['days_employed'] = df['days_employed'].fillna(0) 

Столбец total_income важен для нашего исследования, и заполнение пропусков нулями может исказить статистику. Поэтому заменим их медианным заработком для каждой группы заёмщиков (в данном случае точнее брать значение медианы, а не среднего арифметического, так как экстремально высокие доходы сильно смещают среднее значение по выборке вверх).

In [8]:
df['median_income'] = df.groupby('income_type')['total_income'].transform('median') #вычислим медианный доход по группам

In [9]:
df['total_income'] = df['total_income'].fillna(df['median_income']) 
#заменим пропуски в столбце 'total_income' значениями из столбца 'median_income'

In [10]:
display(df.loc[12], df.loc[41], df.loc[120], df.loc[174]) #проверим несколько строк, где раньше были пропуски

children                           0
days_employed                      0
dob_years                         65
education                    среднее
education_id                       1
family_status       гражданский брак
family_status_id                   1
gender                             M
income_type                пенсионер
debt                               0
total_income                  118514
purpose              сыграть свадьбу
median_income                 118514
Name: 12, dtype: object

children                                           0
days_employed                                      0
dob_years                                         50
education                                    среднее
education_id                                       1
family_status                        женат / замужем
family_status_id                                   0
gender                                             F
income_type                              госслужащий
debt                                               0
total_income                                  150448
purpose             сделка с подержанным автомобилем
median_income                                 150448
Name: 41, dtype: object

children                             0
days_employed                        0
dob_years                           46
education                      среднее
education_id                         1
family_status          женат / замужем
family_status_id                     0
gender                               F
income_type                  сотрудник
debt                                 0
total_income                    142594
purpose             высшее образование
median_income                   142594
Name: 120, dtype: object

children                          0
days_employed                     0
dob_years                        55
education                    высшее
education_id                      0
family_status        вдовец / вдова
family_status_id                  2
gender                            F
income_type               компаньон
debt                              0
total_income                 172358
purpose             свой автомобиль
median_income                172358
Name: 174, dtype: object

In [11]:
df.info() #убедимся, что никаких пропусков не осталось

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
median_income       21525 non-null float64
dtypes: float64(3), int64(5), object(5)
memory usage: 2.1+ MB


#### Вывод

Проанализировав таблицу, мы обнаружили пропуски в столбцах "общий трудовой стаж в днях" и "ежемесячный доход". Мы предполагаем, что некоторые заёмщики по каким-то причинам не стали сообщать величину своего трудового стажа и дохода. Чтобы не потерять важную информацию, содержащуюся в других столбцах таблицы, было решено избавиться от пропущенных значений вызовом метода fillna(). Для простоты мы заменили пропуски в столбце "общий трудовой стаж" нулями. Значение ежемесячного дохода важно для нашего исследования, поэтому пропуски в этом столбце было решено заменить медианным значением, рассчитанным в соответствии с типом дохода для каждого клиента.

### Замена типа данных

Ещё раз обратимся к сводке df.info(). В датафрейме содержатся данные трёх типов: float64(2), int64(5), object(5). Информация строкового типа содержится в столбцах education, family_status, gender, income_type и purpose - содержание ячеек соответствует этому типу, здесь всё в порядке и данные в числа переводить не нужно. Количество дней трудового стажа и доход, которые относятся к типу float64, переведем в данные целочисленного типа, так как для подобной информации точность до шестого знака после запятой не нужна. Для перевода воспользуемся функцией astype().

In [12]:
df['days_employed'] = df['days_employed'].astype(int)

In [13]:
df['total_income'] = df['total_income'].astype(int)

In [14]:
df.head(10) #посмотрим, что получилось

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   Среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   СРЕДНЕЕ             1   
8         2          -6929         35    ВЫСШЕЕ             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  median_income  
0               покупка жилья  142594.396847  
1     приобретение автомобиля  142594.396847  
2               покупка жилья  142594.396847  
3  дополнительное образование  142594.396847  
4             сыграть свадьбу  118514.486412  
5               покупка жилья  172357.950966  
6           операции с жильем  172357.950966  
7                 образование  142594.396847  
8       на проведение свадьбы  142594.396847  
9     покупка жилья для семьи  142594.396847

Теперь данные по стажу и доходам выглядят намного доступнее. Осталось исправить отрицательные значения в столбце days_employed: 

In [15]:
df['days_employed'] = df['days_employed'].abs()

In [16]:
len(df[df['days_employed'] < 0]) #проверим, что строк с отрицательными значениями days_employed не осталось

0

#### Вывод

Чтобы повысить удобство восприятия информации, мы перевели значения столбцов days_employed и income_type из типа float64 (дробные числа, в нашем случае с шестью знаками после запятой) в тип int64 (целые числа). Также мы избавились от отрицательных значений трудового стажа с помощью функции abs().

### Обработка дубликатов

Данные необходимо проверить на дубликаты, которые могли появиться из-за технических ошибок при выгрузке или в самой системе хранения информации о кредитах, например, если заявка от одного человека была ошибочно обработана дважды. Прежде чем искать дубликаты, унифицируем регистр символов в столбце education, чтобы не пропустить никакие повторы.

In [17]:
df['education'] = df['education'].str.lower()

In [18]:
df.duplicated().sum() #посчитаем повторы в таблице

71

In [19]:
df = df.drop_duplicates().reset_index(drop=True) #удалим повторы и обновим индекс

Мы удалили полные повторы, теперь проанализируем информацию более детально. Для ответа на вопросы заказчика нам очень важны данные о детях, семейном положении, уровне дохода и целях кредита. Проверим, есть ли в этих столбцах какие-то артефакты.

In [20]:
df['children'].unique() #проверим, какие уникальные значения встречаются в столбце "количество детей в семье"

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Отрицательное количество детей - в данные явно закралась ошибка, знак минуса тут лишний. Число 20 тоже навевает сомнения: если таких многодетных семей в выгрузке много, можно смело утверждать, что на месте 20 должно быть 2.

In [21]:
df['children'][df['children'] == -1].count() #посчитаем количество семей с минус-одним ребенком

47

In [22]:
df['children'] = df['children'].replace(-1, 1) #заменим отрицательные значения на положительные

In [23]:
df['children'][df['children'] == 20].count() #посчитаем количество семей с двадцатью детьми

76

In [24]:
df['children'] = df['children'].replace(20, 2) #заменим двадцать на два

In [25]:
df['children'].unique() #проверим, что получилось

array([1, 0, 3, 2, 4, 5])

In [26]:
df['family_status'].unique() #с количеством детей разобрались, теперь посмотрим на семейное положение

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [27]:
df['family_status_id'].unique()

array([0, 1, 2, 3, 4])

С семейным положением можно работать. Проверим, всё ли в порядке в столбце "доход". Так как уникальных значений в этом столбце очень много, вместо функции unique() воспользуемся min() и проверим, не встречаются ли там отрицательные значения.

In [28]:
df['total_income'].min()

20667

Кажется, в данных о доходе тоже нет ничего подозрительного. Цели заёмщиков мы изучим ниже, так как для их анализа пригодятся инструменты для работы с естественными языками. На всякий случай ещё раз проверим, что после замен дополнительных дубликатов не появилось. Если их нет, можно переходить к следующему шагу.

In [29]:
df.duplicated().sum()

0

#### Вывод

В целях дальнейшей оптимизации данных мы удалили из таблицы дубликаты с помощью метода drop_duplicates(), а также дополнительно проверили избранные столбцы на предмет артефактов и исправили отрицательное количество детей.

### Лемматизация

Лемматизация пригодится нам для анализа целей заёмщиков. Как мы отмечали ранее, в списке есть похожие формулировки - например, "сыграть свадьбу" и "на проведение свадьбы". Выделение словарных форм поможет нам выяснить, какие цели встречаются наиболее часто.

In [30]:
m = Mystem()

In [31]:
credit_purposes = ', '.join(df['purpose']) #сохраним описание целей в переменной credit_purposes

In [32]:
lemmas = m.lemmatize(credit_purposes) #лемматизируем

In [33]:
Counter(lemmas).most_common() #посчитаем леммы и отсортируем их в порядке убывания

[(' ', 33570),
 (', ', 21453),
 ('недвижимость', 6351),
 ('покупка', 5897),
 ('жилье', 4460),
 ('автомобиль', 4306),
 ('образование', 4013),
 ('с', 2918),
 ('операция', 2604),
 ('свадьба', 2324),
 ('свой', 2230),
 ('на', 2222),
 ('строительство', 1878),
 ('высокий', 1374),
 ('получение', 1314),
 ('коммерческий', 1311),
 ('для', 1289),
 ('жилой', 1230),
 ('сделка', 941),
 ('дополнительный', 906),
 ('заниматься', 904),
 ('подержать', 853),
 ('проведение', 768),
 ('сыграть', 765),
 ('сдача', 651),
 ('семья', 638),
 ('собственный', 635),
 ('со', 627),
 ('ремонт', 607),
 ('приобретение', 461),
 ('профильный', 436),
 ('подержанный', 111),
 ('\n', 1)]

In [34]:
top_purposes = ['недвижимость', 'жилье', 'автомобиль', 'образование', 'свадьба']

#### Вывод

По результатам лемматизации выявлено, что в целях чаще всего встречаются слова "недвижимость", "жилье", "автомобиль", "образование" и "свадьба".

### Категоризация данных

Один из вопросов, поставленных заказчиком, звучит: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?" Проверим, сколько уникальных значений встречается в данных о доходе:

In [35]:
df['total_income'].value_counts()

142594    1070
172357     502
118514     387
150447     145
126262       3
          ... 
101387       1
138249       1
280240       1
390148       1
264193       1
Name: total_income, Length: 18608, dtype: int64

Целых 18608 значений! Работать с ними по отдельности будет неудобно, поэтому нужно выделить какие-то категории.

In [36]:
df['total_income'].mean() #найдем среднее значение ежемесячного дохода для всей выборки

165319.57229421087

In [37]:
df['total_income'].median() #найдем медианное значение ежемесячного дохода для всей выборки

142594.0

Средний ежемесячный доход по выборке составляет 165 тысяч, медианный - 142,5 тысячи. Поскольку значение среднего арифметического заметно смещено вверх, для дальнейшего анализа будем использовать медиану в качестве среднего дохода. Чтобы определить пороги низкого и высокого дохода, вычислим значение для 20% и 80% выборки.

In [38]:
df['total_income'].quantile([.2, .8]) 

0.2     98537.6
0.8    214618.2
Name: total_income, dtype: float64

Выделим условные группы следующим образом: менее 98537 - низкий доход, от 98537 до 142594 - доход ниже среднего, от 142594 до 214618 - доход выше среднего, более 214618 - высокий доход. Напишем функцию для распределения заёмщиков по категориям:

In [39]:
def def_income_level(total_income):
    if total_income < 98537:
        return 'низкий доход'
    if 98537 <= total_income < 142594:
        return 'ниже среднего'
    if 142594 <= total_income < 214618:
        return 'выше среднего'
    else:
        return 'высокий доход'

In [40]:
df['income_level'] = df['total_income'].apply(def_income_level)

Мы добавили в таблицу новый столбец income_level с информацией о уровне дохода. Аналогично разделим заёмщиков на две большие категории - с детьми и бездетных. Для ответа на вопрос заказчика "Есть ли зависимость между наличием детей и возвратом кредита в срок?" количество детей нам не так важно, а с двумя категориями удобнее работать. 

In [41]:
def children_status(count):
    if count > 0:
        return 'детный'
    else:
        return 'бездетный'

In [42]:
df['children_status'] = df['children'].apply(children_status)

Также будет полезно распределить цели клиентов по категориям. Для этого обратимся к результатам проведённой лемматизации и выделим самые популярные причины для займов.

In [43]:
def def_purpose_category(row):
    for lemma in row:
        for purpose in top_purposes:
            if lemma == purpose:
                category = lemma
    return category

In [44]:
df['purpose'] = df['purpose'].apply(lambda x: m.lemmatize(x))

In [45]:
df['purpose_category'] = df['purpose'].apply(def_purpose_category)

In [46]:
df.sample(7) #проверим, что получилось

children  days_employed  dob_years education  education_id  \
4629          0         367101         63   среднее             1   
12546         0           1673         59   среднее             1   
5964          0           1680         28   среднее             1   
18529         0         342085         53   среднее             1   
5704          0           2363         59   среднее             1   
15340         1              0         30   среднее             1   
6528          0           2502         42    высшее             0   

         family_status  family_status_id gender income_type  debt  \
4629   женат / замужем                 0      M   пенсионер     0   
12546  женат / замужем                 0      M   сотрудник     0   
5964   женат / замужем                 0      F   сотрудник     0   
18529  женат / замужем                 0      F   пенсионер     0   
5704   женат / замужем                 0      M   сотрудник     0   
15340  женат / замужем                 0      M   сотрудник     0   
6528   женат / замужем                 0      F   компаньон     0   

       total_income                                         purpose  \
4629         127745                   [высокий,  , образование, \n]   
12546        134519             [на,  , покупка,  , автомобиль, \n]   
5964          80719                                     [жилье, \n]   
18529         80130  [строительство,  , жилой,  , недвижимость, \n]   
5704         165470                 [получение,  , образование, \n]   
15340        142594            [дополнительный,  , образование, \n]   
6528         242327                          [ремонт,  , жилье, \n]   

       median_income   income_level children_status purpose_category  
4629   118514.486412  ниже среднего       бездетный      образование  
12546  142594.396847  ниже среднего       бездетный       автомобиль  
5964   142594.396847   низкий доход       бездетный            жилье  
18529  118514.486412   низкий доход       бездетный     недвижимость  
5704   142594.396847  выше среднего       бездетный      образование  
15340  142594.396847  выше среднего          детный      образование  
6528   172357.950966  высокий доход       бездетный            жилье

#### Вывод

В соответствии с целями нашего исследования мы распределили клиентов по следующим категориям:
1. Четыре категории по уровню дохода: высокий, выше среднего, ниже среднего и низкий;
2. Две категории по наличию детей: детный и бездетный;
3. Пять категорий по цели займа: недвижимость (в т.ч. коммерческая), жильё, автомобиль, образование, свадьба.

## Ответы на вопросы

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

Информация по возврату кредита в срок содержится в столбце "debt": значение равно единице, если клиент имеет задолженность по возврату кредитов, или нулю, если этого за ним не замечено. Соберём данные по количеству детей и задолженностям в сводную таблицу и проанализируем данные.

In [47]:
child_pivot = df.pivot_table(index='children_status', values='debt', aggfunc=['count', 'sum']) 
#count считает общее число займов в каждой категории, sum - количество случаев невозврата кредита

In [48]:
child_pivot['ratio'] = child_pivot['sum'] / child_pivot['count'] * 100 #отношение случаев невозврата к общему числу займов

In [49]:
display(child_pivot)

count   sum     ratio
                  debt  debt          
children_status                       
бездетный        14091  1063  7.543822
детный            7363   678  9.208203

Судя по имеющейся информации, бездетные люди берут больше кредитов и в абсолютных числах имеют больше долгов, однако соотношение задолженностей к общему числу кредитов говорит об обратном: заёмщики с детьми чаще не возвращают кредиты вовремя.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [50]:
family_pivot = df.pivot_table(index='family_status', values='debt', aggfunc=['count', 'sum']) 

In [51]:
family_pivot['ratio'] = family_pivot['sum'] / family_pivot['count'] * 100

In [52]:
display(family_pivot)

count  sum     ratio
                        debt debt          
family_status                              
Не женат / не замужем   2810  274  9.750890
в разводе               1195   85  7.112971
вдовец / вдова           959   63  6.569343
гражданский брак        4151  388  9.347145
женат / замужем        12339  931  7.545182

Наиболее часто задолженности по кредитам возникают у неженатых клиентов и незамужних клиенток, а также тех, кто состоит в гражданском браке. Интересно, что разведённые и овдовевшие возвращают кредиты даже более аккуратно, чем состоящие в браке, поэтому при прочих равных мы можем рекомендовать банку рассматривать факт наличия штампа в паспорте как дополнительный фактор надёжности заёмщика, даже если его брак по каким-то причинам уже прекратился.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [53]:
money_pivot = df.pivot_table(index='income_level', values='debt', aggfunc=['count', 'sum']) 

In [54]:
money_pivot['ratio'] = money_pivot['sum'] / money_pivot['count'] * 100

In [55]:
display(money_pivot)

count  sum     ratio
               debt debt          
income_level                      
высокий доход  4291  300  6.991377
выше среднего  7390  629  8.511502
ниже среднего  5482  468  8.537030
низкий доход   4291  344  8.016779

Задолженности чаще возникают у людей со средними и низкими доходами, тогда как клиенты с высоким доходом (более 215 тысяч в месяц) в основном возвращают кредиты вовремя.

### Как разные цели кредита влияют на его возврат в срок?

In [56]:
purposes_pivot = df.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'sum']) 

In [57]:
purposes_pivot['ratio'] = purposes_pivot['sum'] / purposes_pivot['count'] * 100

In [58]:
display(purposes_pivot)

count  sum     ratio
                  debt debt          
purpose_category                     
автомобиль        4306  403  9.359034
жилье             4460  308  6.905830
недвижимость      6351  474  7.463392
образование       4013  370  9.220035
свадьба           2324  186  8.003442

Наиболее часто задолженности возникают у клиентов, берущих средства на покупку автомобиля, на втором месте - образовательные кредиты. Низкий процент долгов по кредитам с целью покупки жилья можно объяснить, с одной стороны, тем, что клиенты более аккуратно рассчитывают свои возможности по выплате, собираясь взять долгосрочный займ, с другой стороны, можно предположить, что банк со своей стороны тоже более внимательно отбирает заявки кандидатов на ипотеку.

## Общий вывод

Проанализировав предоставленную статистику, мы можем утверждать, что наличие детей, семейное положение и уровень дохода оказывают прямое влияние на платёжеспособность клиентов:
- Клиенты, находящиеся в браке, реже опаздывают с погашением кредита: вероятно, доход супруга служит в этом случае дополнительной страховкой. 
- Наличие детей, напротив, оказывает негативное влияние: очевидно, что воспитание детей требует значительных средств и сопряжено с неожиданными затратами, поэтому по сравнению с бездетными клиенты с детьми чаще задерживают выплаты по кредитам. 
- Клиенты с высоким уровнем ежемесячного дохода более аккуратно возвращают кредиты, чем клиенты со средними и низкими доходами.

Стоит отметить, что в полученных данных по доходу мы обнаружили большое количество пропусков, что могло повлиять на результаты исследования. Данный пункт заслуживает внимания: если пропуски возникли не из-за технической ошибки при выгрузке, а из-за того, что при подаче заявки на кредит указание дохода не является обязательным, мы рекомендуем банку сделать его обязательным, чтобы получать более полную картину для кредитного скоринга.

Изучив влияние целей кредита на его своевременный возврат, мы выяснили, что самыми аккуратными плательщиками являются клиенты, взявшие кредит на покупку жилья. Таким образом, на основе имеющейся информации идеальным клиентом банка являются молодые семьи без детей с высоким доходом, желающие взять ипотеку, однако предлагаем ими не ограничиваться. Также мы рекомендуем банку сделать в форме заявки выпадающий список целей для удобства его автоматической обработки.